routing/datasource.go

```diff
@@ -608,9 +612,15 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
 			invalidRouteIds := make(map[string]struct{})
 			validRoutes := []*eskip.Route{}
 
-			for _, err := range errs {
-				o.Log.Error(err)
-				invalidRouteIds[err.ID] = struct{}{}
+			if len(errs) > 0 {
+				for _, err := range errs {
+					o.Log.Error(err)
+					invalidRouteIds[err.ID] = struct{}{}
+				}
+
+				if o.Metrics != nil {
+					o.Metrics.UpdateInvalidRoute(map[string]int{"matcher_error": len(errs)})
```

Comment:
Lets define `errInvalidMatcher` to have this all unified and use like:
```go
var errInvalidMatcher         = invalidDefinitionError("invalid_matcher")
...
if o.Metrics != nil && len(errs) > 0 {
    o.Metrics.UpdateInvalidRoute(map[string]int{errInvalidMatcher.Code(): len(errs)})
}
```

In theory we could also add `Unwrap` to the `definitionError` and then use `invalidDefinitionError` wrapper for `Original` to have distinct errors within `newMatcher` but lets keep it simple for now.

Reply:
Thinking about `len(errs) > 0` condition I think we have an edge case here and in processRoutesDefs that we do not reset gauges to zero. 

So if value for a certain error code was not zero and then routes were fixed we should set value to zero. 
Here we can remove `len(errs) > 0` and in processRoutesDefs set zeroes explicitly for all defined error codes.